route: Add support of quickack

cathay4t opened the pull request:
Support quickack property of route, for example:

```yml
routes:
  config:
    - destination: 203.0.113.0/24
      next-hop-address: 192.0.2.1
      next-hop-interface: eth1
      quickack: true
    - destination: 2001:db8:c::/64
      next-hop-address: 2001:db8:1::2
      next-hop-interface: eth1
      quickack: false
```

Unit and integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-80418

cathay4t committed:
route: Add gen_conf support for initrwnd,initcwnd and mtu

Integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
The failure of nm_stable on slow group is known issue of NetworkManager-1.51.90-2.el9 which should be fixed by pending update of NM.

liangwen12year commented on rust/src/lib/nispor/route.rs:
```diff
@@ -209,6 +209,7 @@ fn np_route_to_nmstate(np_route: &nispor::Route) -> RouteEntry {
     route_entry.initcwnd = np_route.initcwnd;
     route_entry.initrwnd = np_route.initrwnd;
     route_entry.mtu = np_route.mtu;
+    route_entry.quickack = np_route.quickack.map(|q| q > 0);
```
If the quickack only meaningful when the quickack is true? Why do we always omit quickack false when showing the property?

cathay4t replied:
default is `quickack: false`.

liangwen12year reviewed (commented):
I applied the same route config as provided in the commit message the first time,  I noticed that this route appeared in the NetworkManager connection profile: 
```
ipv6.routes:                            { ip = 2001:db8:c::/64, nh = 2001:db8:1::2, mt = 1024 quickack=false table=254 }
``` 
but when we applied the same config multiple times, the `quickack=false` did not appear in the connection profile anymore: 
```
ipv6.routes:                            { ip = 2001:db8:c::/64, nh = 2001:db8:1::2, mt = 1024 table=254 }
```

Is this expected? maybe we should still keep the ` quickack=false` in the connection profile when applying multiple times?

cathay4t commented:
> I applied the same route config as provided in the commit message the first time, I noticed that this route appeared in the NetworkManager connection profile:
> 
> ```
> ipv6.routes: { ip = 2001:db8:c::/64, nh = 2001:db8:1::2, mt = 1024 quickack=false table=254 }
> ```
> 
> but when we applied the same config multiple times, the `quickack=false` did not appear in the connection profile anymore:
> 
> ```
> ipv6.routes:  { ip = 2001:db8:c::/64, nh = 2001:db8:1::2, mt = 1024 table=254 }
> ```
> 
> Is this expected? maybe we should still keep the ` quickack=false` in the connection profile when applying multiple times?

The `quickack=false` == `no quickack`. I don't see a problem on this.

ihuguet commented on rust/src/lib/query_apply/route.rs:
```diff
@@ -74,16 +74,17 @@ impl MergedRoutes {
                         continue;
                     }
                 }
+                cur_rt.sanitize_current_route_for_verify();
```
Why is this needed for quickack and not for other options?

cathay4t replied:
Because no other route property is bool.

ihuguet replied:
Can you expand a bit more? How does that affect us?

cathay4t replied:
When user desire `quickack: false`, during verification stage, the query output will not have `quaickack` because `quickack: false` mean no flag. If we do not sanitize current for verify, we will get a verification error there.

ihuguet replied:
Can you add a comment explaining it?
Apart from that, LGTM and can be merged.

cathay4t replied:
Comment added.

ihuguet commented on rust/src/lib/route.rs:
```diff
@@ -264,6 +264,13 @@ pub struct RouteEntry {
         deserialize_with = "crate::deserializer::option_u32_or_string"
     )]
     pub mtu: Option<u32>,
+    /// MTU
```
Wrong comment

cathay4t replied:
Fixed.

cathay4t committed:
route: Add support of quickack

Support quickack property of route, for example:

```yml
routes:
  config:
    - destination: 203.0.113.0/24
      next-hop-address: 192.0.2.1
      next-hop-interface: eth1
      quickack: true
    - destination: 2001:db8:c::/64
      next-hop-address: 2001:db8:1::2
      next-hop-interface: eth1
      quickack: false
```

Unit and integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-80418

Signed-off-by: Gris Ge <fge@redhat.com>
